.github/workflows/release.yml

```diff
@@ -0,0 +1,99 @@
+
+jobs:
+  release:
+    runs-on: ubuntu-latest
+    steps:
+      - run: |
+          if [[ $GITHUB_REF_NAME != release/* ]]; then
+            echo the release workflow should only be run against release branches
+            exit 1
+          fi
+
+      - uses: actions/checkout@v3
+
+      - name: Set environment variables
+        run: |
+          stable_version=$(./scripts/eachdist.py version --mode stable)
+          prerelease_version=$(./scripts/eachdist.py version --mode prerelease)
+
+          if [[ $stable_version =~ ([0-9]+)\.([0-9]+)\.([0-9]+) ]]; then
+            stable_major="${BASH_REMATCH[1]}"
+            stable_minor="${BASH_REMATCH[2]}"
+            stable_patch="${BASH_REMATCH[3]}"
+          else
+            echo "unexpected stable_version: $stable_version"
+            exit 1
+          fi
+          if [[ $stable_patch != 0 ]]; then
+            if [[ $prerelease_version =~ 0\.([0-9]+)b([0-9]+) ]]; then
+              prerelease_minor="${BASH_REMATCH[1]}"
+              prerelease_patch="${BASH_REMATCH[2]}"
+            else
+              echo "unexpected prerelease_version: $prerelease_version"
+              exit 1
+            fi
+            if [[ $prerelease_patch != 0 ]]; then
+              prior_combined_version_when_patch="$stable_major.$stable_minor.$((stable_patch - 1))/0.${prerelease_minor}b$((prerelease_patch - 1))"
+            fi
+          fi
+
+          echo "STABLE_VERSION=$stable_version" >> $GITHUB_ENV
+          echo "PRERELEASE_VERSION=$prerelease_version" >> $GITHUB_ENV
+
+          echo "PRIOR_COMBINED_VERSION_WHEN_PATCH=$prior_combined_version_when_patch" >> $GITHUB_ENV
+
+      - name: Generate release notes
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          # conditional block not indented because of the heredoc
+          if [[ ! -z $PRIOR_COMBINED_VERSION_WHEN_PATCH ]]; then
+          cat > release-notes.txt << EOF
+          This is a patch release on the previous $PRIOR_COMBINED_VERSION_WHEN_PATCH release, fixing the issue(s) below.
+
+          EOF
+          fi
+
+          # the complex perl regex is needed because markdown docs render newlines as soft wraps
+          # while release notes render them as line breaks
+          sed -n "0,/^## Version $STABLE_VERSION\/$PRERELEASE_VERSION/d;/^## Version /q;p" CHANGELOG.md \
+            | perl -0pe 's/(?<!\n)\n *(?!\n)(?![-*] )(?![1-9]+\. )/ /g' \
+            >> release-notes.txt
+
+      - name: Create GitHub release
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh release create --target $GITHUB_REF_NAME \
+                            --title "Version $STABLE_VERSION/$PRERELEASE_VERSION" \
+                            --notes-file release-notes.txt \
+                            --discussion-category announcements \
+                            v$STABLE_VERSION
+
+      - name: Update the change log with the release date
```

Comment:
Curious as to why the change log is updated with the release date AFTER the release has already been created? Don't we want to have the published package to have the release date reflected in the CHANGELOG?

Reply:
The release date could be populated in the "prepare release branch" PR. The only downside of that is the actual release date might be later.

Maybe do both? Populate it in the "prepare release branch" PR, and in the post-release workflow, update the date if it happened to be later? Or maybe even fail to release if the release date doesn't match today's date, and have it push a PR to update the date at that point (before the release, so that the release date is always correct +/- small error in the release)?

Reply:
> Populate it in the "prepare release branch" PR, and in the post-release workflow, update the date if it happened to be later?

I think this would be good. As for preventing the release, I think small +/- error is not a big deal.

Reply:
ya, I like this, I updated the template-repository and this PR to update the date in the "prepare release branch" and "prepare patch release" workflows as part of the existing version bumping PRs